### ISSUE TITLE
Investigate using Remote-SSH's local server in Remote-Containers

### ISSUE THREAD

**Issue by chrmarti** [microsoft/vscode-remote-release#6218]
From https://github.com/microsoft/vscode-remote-release/issues/6137#issuecomment-1017985665.

Options for API:
- Remote-SSH could export API from its `activate()` function. It is not clear if a remote resolver can use another extension's API before resolving has finished. /cc @alexdima 
- Remote-SSH could place a module in its extension folder where Remote-Containers can load it from. (Remote-Containers can lookup Remote-SSH's extension folder using VS Code's extension API.)

**Comment by alexdima** [comment]
> It is not clear if a remote resolver can use another extension's API before resolving has finished

Yes, it is possible to depend on another extension from a resolver and use the API provided during resolve time.

But wouldn't this mean that `Remote-Containers` needs to take an extension dependency on `Remote-SSH`? It might be architecturally better if code sharing is done via a node module.

**Comment by chrmarti** [comment]
@alexdima Can we use the extension's API without depending on it? The extension API would allow for activating it, but I forgot the precise restrictions there.

**Comment by alexdima** [comment]
Yes, it is possible to use `vscode.extensions.getExtension` to probe if the other extension is available. Have you considered code sharing via a node module?

**Comment by roblourens** [comment]
I would rather not share code with npm if we can avoid it just because it complicates things for us. It doesn't seem that bad to have an extension dependency on Remote-SSH, when you have features that explicitly integrate with Remote-SSH.

**Comment by alexdima** [comment]
:+1: I thought the current relationship is that the two extensions integrate with each-other but not depend on each other. With this proposal would Remote-Containers continue to work when Remote-SSH is not installed or disabled?

**Comment by roblourens** [comment]
I guess that since this feature isn't active unless you are already using Remote-SSH anyway, using the API without having an extension dependency is fine.

**Comment by alexdima** [comment]
> I guess that since this feature isn't active unless you are already using Remote-SSH anyway, using the API without having an extension dependency is fine.

:+1: Aha! It wasn't clear to me that this reuse would occur only when using Remote-Containers and Remote-SSH together. I thought "local server" means the code for the local server proxy created by dev containers and which is returned as a resolver result in all cases.

**Comment by chrmarti** [comment]
"local server" is a server for sharing SSH connections that Remote-SSH has. It sounds like we can use `vscode.extensions.getExtension` to detect when Remote-SSH is available and either activate it to get at API or we could load a module from the Remote-SSH's install folder if that's cleaner or more efficient.

An NPM package would allow Remote-Containers to reuse the local server in the absence of Remote-SSH, but I think that might be a rare case (a user using SSH for dev containers without installing Remote-SSH).

**Comment by alexdima** [comment]
:+1: Thanks for the extra info! I agree that `vscode.extensions.getExtension`, activating it if it's there and using its API sounds like the cleanest approach.

**Comment by roblourens** [comment]
Is this still relevant/interesting @chrmarti?

**Comment by chrmarti** [comment]
It would help Dev Containers avoid asking for passwords / passphrases when there already is a ssh connection and it would reuse the handling of edge cases.
